src/nnpack/interface.jl

```diff
@@ -0,0 +1,143 @@
+function check_support(x, k, pad, stride, dilation = 1)
+    fallback = false
+    dilation == 1 || dilation == (1, 1) || (fallback = true)
+    pad_, stride_ = expand(Val{length(k)}, pad), expand(Val{length(k)}, stride)
+    ((size(x, 1) - k[1] + 2 * pad_[1]) % stride_[1] == 0 && (size(x, 2) - k[2] + 2 * pad_[2]) % stride_[2] == 0) || (fallback = true)
+    return pad_, stride_, fallback
+end
+
+#NOTE: Commenting out the activation functions until sure what to do
+
+# relu(x::AA1) = nnp_relu_output(x, inplace ? x : similar(x), threadpool = shared_threadpool[])
+
+# leakyrelu(x::AA1, a = oftype(x/1, 0.01)) =
+#     nnp_relu_output(x, inplace ? x : similar(x), negative_slope = a, threadpool = shared_threadpool[])
+
+softmax!(x::A) where A<:AbstractVecOrMat{Float64} = softmax!(Float32.(x))
```

Comment:
this seems very wrong for 3 reasons:
1) we are not modifying x as `softmax!(x)` is supposed to do; 
2) If I choose to use Float64 arrays you have to respect that, I don't want things
to be rounded to 32 bits internally without me (the user) knowing it. 
3) also, the return type is changing to float32 array

Reply:
There was a prior discussion on the internal conversion issue.

Here's @staticfloat 's comment on the issue

> If we're calling into NNPACK, I think it makes sense to "poison" the chain of operations; e.g. if we've configured Flux to be dead-set on using NNPACK, then conv(X<:Float64, W<:Float32) should return a Float32; narrowing the output type due to the internal kernel working in f32. This should cause downstream operations to use conv(X<:Float32, W<:Float32), thereby avoiding the constant re-quantization.
If we just document this, and allow people who really care about precision to trivially re-cast to Float64 in the places that it matters, then I think we should be able to get the best of both worlds.

However, we are currently making any `Float64` conversion to `Float32` to proceed. This speeds up the training by a factor of 4x. So I believe its worth making the conversion and mentioning about it in the documentation.

Also for the first point, `softmax!` defined for `Float32` will be making the inplace modification of x. And we are calling that from the function for `Float64`. Were you refering to this?

Reply:
>If we're calling into NNPACK, I think it makes sense to "poison" the chain of operations; e.g. if we've configured Flux to be dead-set on using NNPACK, then conv(X<:Float64, W<:Float32) should return a Float32; narrowing the output type due to the internal kernel working in f32. This should cause downstream operations to use conv(X<:Float32, W<:Float32), thereby avoiding the constant re-quantization.
If we just document this, and allow people who really care about precision to trivially re-cast to Float64 in the places that it matters, then I think we should be able to get the best of both worlds.

I disagree with that, we should follow julia promotion conventions and have
`conv(X<:Float64, W<:Float32) -> Float64`. 

If one wants to work only in float32, he would simply have to cast the input X to float32 and every operation in the successive layers will be float32.

On the contrary, having `conv(X<:Float64, W<:Float32) -> Float32`, beside being surprising to the user, would mean that if I want the "work with 64 bit activation" behaviour, I would have to cast
to float64 at *each* hidden layer. 

And here we are discussing a Float64 x Float32 -> Float32 operation, a Float64 -> Float32 operation as the proposed `softmax` makes even less sense. If NNPACK doesn't support float64 operations (is this the case?) we have to fall back to the julia implementation for float64 ones. 

We should not enforce something on the users, they should be in power, if the want speed they should use float32, but if they precision it has to possible for them to stick to float64.

>Also for the first point, softmax! defined for Float32 will be making the inplace modification of x. And we are calling that from the function for Float64. Were you refering to this?

Problem is that when you call `Float32.(x)` you are creating a new array, so `x` is not modified at all.

Reply:
NNPACK does not support float64 operations, as far as I know, and this PR was (in my understanding) code that only gets enabled when Flux is in an NNPACK-specific mode, as stated in the block of text you quoted above.  If that's not the case then I agree we need to change something, but if Flux is setup to use NNPACK as its backing execution engine, this behavior makes sense.

Reply:
Currently, NNPACK gets called by default as long as the hardware supports it. But we can simply add a flag which can make us use the NNlib default functions instead of NNPACK. I believe if the user is allowing NNPACK to load then there should not be any issue if we internally do computation in Float32 because NNPACK cant handle Float64.

Also `conv(X<:Float64, W<:Float32)` will still error in this PR. The datatypes of the arrays must match for the convolution to take place. I could try to make the changes here itself. But it also involves making a lot of changes in base nnlib functions. Hence if you do `conv(X<:Float64, W<:Float64)` we get `Float32` outputs in case you enable NNPACK.

If accurary is an issue and still user wants the speed of NNPACK, we might want the user to pass another flag which loads the NNPACK `Float32` functions but restricts the usage of NNPACK for `Float64`.

> Problem is that when you call Float32.(x) you are creating a new array, so x is not modified at all.

Yes, you are right. I will correct this.